cli/src/action/database/upgrade/scabbard.rs

```diff
@@ -0,0 +1,231 @@
+            ConnectionUri::Postgres(url) => {
+                let connection_manager = ConnectionManager::<diesel::pg::PgConnection>::new(url);
+                let pool = Pool::builder().build(connection_manager).map_err(|err| {
+                    InternalError::from_source_with_prefix(
+                        Box::new(err),
+                        "Failed to build connection pool".to_string(),
+                    )
+                })?;
+                // Test the connection
+                let _conn = pool
+                    .get()
+                    .map_err(|err| InternalError::from_source(Box::new(err)))?;
+                Ok(Self {
+                    admin_service_store: Box::new(DieselAdminServiceStore::new(pool.clone())),
+                    commit_hash_store_factory: CommitHashStoreFactory::Postgres(pool.clone()),
+                    node_id_store: Box::new(DieselNodeIdStore::new(pool)),
+                })
+            }
+            #[cfg(feature = "sqlite")]
+            ConnectionUri::Sqlite(conn_str) => {
+                if (conn_str != ":memory:") && !std::path::Path::new(&conn_str).exists() {
+                    return Err(UpgradeError::Internal(InternalError::with_message(
+                        format!("Database file '{}' does not exist", conn_str),
+                    )));
+                }
+                let connection_manager =
+                    ConnectionManager::<diesel::sqlite::SqliteConnection>::new(conn_str);
+                let mut pool_builder = Pool::builder();
+                // A new database is created for each connection to the in-memory SQLite
+                // implementation; to ensure that the resulting stores will operate on the same
+                // database, only one connection is allowed.
+                if conn_str == ":memory:" {
+                    pool_builder = pool_builder.max_size(1);
+                }
+                let pool = pool_builder.build(connection_manager).map_err(|err| {
+                    InternalError::from_source_with_prefix(
+                        Box::new(err),
+                        "Failed to build connection pool".to_string(),
+                    )
+                })?;
+                // Test the connection
+                let _conn = pool
+                    .get()
+                    .map_err(|err| InternalError::from_source(Box::new(err)))?;
+                Ok(Self {
+                    admin_service_store: Box::new(DieselAdminServiceStore::new(pool.clone())),
+                    commit_hash_store_factory: CommitHashStoreFactory::Sqlite(pool.clone()),
+                    node_id_store: Box::new(DieselNodeIdStore::new(pool)),
+                })
+            }
+            #[cfg(feature = "memory")]
+            _ => Err(UpgradeError::InvalidState(InvalidStateError::with_message(
+                format!("Unsupported database: {}", database_uri),
+            ))),
+        }
+    }
+}
+
+enum CommitHashStoreFactory {
```

Comment:
This should be implemented in a way which provides the benefit of abstraction/encapsulation. Currently, having this factory enum provides no opportunity for the code using it to not be postgres/sqlite-aware. If it were a trait on two different implementations, then only the creation of the factory would need to be postgres/sqlite-aware.

Reply:
I guess maybe some of the benefits of the abstraction are here, and the biggest issues would be a) the abstraction can be circumvented and the pool accessed directly (thus making the caller postgres/sqlite-aware); b) with this pattern, the implementations are fixed (you could not write another of these factory implementations like you can an abstract factory), which is perhaps ok here but wouldn't be in other places.

Reply:
I've replaced this with a similar pattern to libsplinter.